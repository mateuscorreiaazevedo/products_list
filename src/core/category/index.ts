export * from './entities/category.entity'
export * from './contracts/category.contract'
export * from './services/http/http-client-category.service'
export * from './mappers/category.mapper'
export * from './dtos/list-categories.dto'
export * from './services/mock/mock-category.service'
export * from './usecases/list-categories.usecase'
export * from './factories/category.factory'
