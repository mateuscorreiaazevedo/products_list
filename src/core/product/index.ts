export * from './services/http/http-client-product.service'
export * from './usecases/list-products.usecase'
export * from './contracts/products.contract'
export * from './factories/product.factory'
export * from './entities/product.entity'
