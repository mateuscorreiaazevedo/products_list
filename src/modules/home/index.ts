export * from './components/category-select-filter'
export * from './components/product-item-card'

export * from './hooks/set-filter-products-by-category'
export * from './hooks/list-products-reducer'
export * from './hooks/list-products'
