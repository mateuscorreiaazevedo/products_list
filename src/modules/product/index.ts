export * from './components/product-item-card'
export * from './hooks/list-products-reducer'
export * from './hooks/list-products'
