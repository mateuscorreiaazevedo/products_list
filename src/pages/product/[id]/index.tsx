import { type FindProductByIdResponseDTO, ProductFactory } from '@/core/product'
import { ProductBanner } from '@/modules/product'
import { HStack, Text, VStack } from '@/shared/components'
import { useTheme } from '@/shared/contexts'
import type { GetServerSideProps } from 'next'
import Head from 'next/head'
import { darken, lighten } from 'polished'

export const getServerSideProps: GetServerSideProps = async ({ params }) => {
  const id = params?.id

  const productById = await ProductFactory.http().findProductById({ id: Number(id) })

  return {
    props: {
      ...productById,
    },
  }
}

export default function ProductItem({ data }: FindProductByIdResponseDTO) {
  const { theme } = useTheme()

  return (
    <>
      <Head>
        <title>
          Product {data.id} - {data.model}
        </title>
      </Head>
      <VStack width={theme.spacing.full} gap={theme.spacing[5]}>
        <HStack width={theme.spacing.full}>
          <Text
            as="h1"
            font={theme.text.heading}
            color={
              theme.title === 'light'
                ? lighten(0.2, theme.colors.foreground)
                : darken(0.2, theme.colors.foreground)
            }
          >
            {data.title}
          </Text>
        </HStack>
        <ProductBanner {...data} />
      </VStack>
    </>
  )
}
