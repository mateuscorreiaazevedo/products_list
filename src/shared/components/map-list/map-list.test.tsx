import { CategoryFactory } from '@/core/category'
import { ProductFactory } from '@/core/product'
import { ProductItemCard } from '@/modules/home'
import type { ComponentType } from 'react'
import { MapList } from '.'

const categoriesMock = async () => await CategoryFactory.mock().listCategories()
const productsMock = async () => await ProductFactory.mock().listProducts()

function renderComponent<T = any>(
  data: T[],
  item: ComponentType<{
    item: T
    index: number
  }>,
) {
  return <MapList data={data} renderItem={item} />
}

describe('MapList', () => {
  test('Should render a list of text', async () => {
    const { data } = await categoriesMock()
    const result = renderComponent(data, ({ item }) => (
      <span key={item.value}>{item.label}</span>
    ))

    expect(result).toMatchSnapshot()
    expect(result.props.data).toEqual(data)
  })

  test('Should render a list of custom components', async () => {
    const { data } = await productsMock()
    const response = renderComponent(data, ({ item }) => (
      <ProductItemCard {...item} key={item.id} />
    ))

    expect(response).toMatchSnapshot()
    expect(response.props.data).toEqual(data)
    expect(response.props.renderItem).toBeInstanceOf(Function)
  })
})
