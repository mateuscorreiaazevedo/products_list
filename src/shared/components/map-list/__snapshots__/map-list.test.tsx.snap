// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapList Should render a list of custom components 1`] = `
{
  "$$typeof": Symbol(react.transitional.element),
  "_owner": null,
  "_store": {},
  "key": null,
  "props": {
    "data": [
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 1.",
        "id": 1,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 1",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 2.",
        "id": 2,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 2",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 3.",
        "id": 3,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 3",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 4.",
        "id": 4,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 4",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 5.",
        "id": 5,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 5",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 6.",
        "id": 6,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 6",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 7.",
        "id": 7,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 7",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 8.",
        "id": 8,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 8",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 9.",
        "id": 9,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 9",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 10.",
        "id": 10,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 10",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 11.",
        "id": 11,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 11",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 12.",
        "id": 12,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 12",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 13.",
        "id": 13,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 13",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 14.",
        "id": 14,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 14",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 15.",
        "id": 15,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 15",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 16.",
        "id": 16,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 16",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 17.",
        "id": 17,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 17",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 18.",
        "id": 18,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 18",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 19.",
        "id": 19,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 19",
      },
      {
        "brand": "Marca Exemplo",
        "description": "Descrição do produto 20.",
        "id": 20,
        "image": "",
        "model": "Modelo Exemplo",
        "price": "$19.99",
        "title": "Produto 20",
      },
    ],
    "renderItem": [Function],
  },
  "type": [Function],
}
`;

exports[`MapList Should render a list of text 1`] = `
{
  "$$typeof": Symbol(react.transitional.element),
  "_owner": null,
  "_store": {},
  "key": null,
  "props": {
    "data": [
      {
        "label": "Tv",
        "value": "tv",
      },
      {
        "label": "Audio",
        "value": "audio",
      },
      {
        "label": "Laptop",
        "value": "laptop",
      },
      {
        "label": "Mobile",
        "value": "mobile",
      },
      {
        "label": "Gaming",
        "value": "gaming",
      },
      {
        "label": "Appliances",
        "value": "appliances",
      },
    ],
    "renderItem": [Function],
  },
  "type": [Function],
}
`;
